readers-frontend/src/pages/[year]/[month]/[day]/[slug].astro

```diff
@@ -0,0 +1,326 @@
+---
+import Article from "@components/Article.astro";
+import ArticleCookieManager from "@components/ArticleCookieManager.astro";
+import TopArticles from "@components/SolidJS/TopArticles";
+import {
+  articleWithUserAndDraft,
+  articlesWithArticleDrafts,
+  getAllArticles,
+  getArticleImage,
+  getArticles,
+} from "@components/utils/articles";
+import type { articlesCategories, categories } from "@prisma/client";
+import { Icon } from "astro-icon";
+import Layout from "src/layouts/Layout.astro";
+import prisma from "src/prisma";
+
+type ArticleCategories = articlesCategories & {
+  category: categories;
+};
+
+export const getParentCategory = (categories: ArticleCategories[]) => {
+  // I think categories commonly come from the backend as:
+  // Top - Middle - Bottom for examples. Where Bottom is nested
+  // under Middle and Top is an "extra category". Hence, the reverse,
+  // trying to find related ones.
+  const interimParentCategory = categories
+    .slice()
+    .reverse()
+    .find(
+      ({ category }) =>
+        (category.categories_nestUnder === null && // Nouse
+          category.categories_id !== 4) ||
+        category.categories_nestUnder === 4 // Muse
+    );
+
+  if (interimParentCategory) {
+    if (
+      // This is mainly for Muse articles as they are
+      // always going to have the Muse category listed
+      categories.length < 3 ||
+      categories.find(
+        ({ category }) =>
+          category.categories_nestUnder ===
+          interimParentCategory.category.categories_id
+      )
+    )
+      return interimParentCategory.category;
+  }
+
+  return categories[0].category;
+};
+
+export const dateFormatter = new Intl.DateTimeFormat("en-GB", {
+  year: "numeric",
+  month: "2-digit",
+  day: "2-digit",
+});
+
+export async function getStaticPaths() {
+  const articles = (await getAllArticles()).filter((article) => {
+    // Ensure article has an articleDraft
+    if (article.articlesDrafts.length === 0) return false;
+
+    // Ensure article has categories
+    if (article.categories.length === 0) return false;
+
+    // Some articles' slug is not formatted properly
+    const slug = article.articles_slug;
+    if (slug)
+      return (
+        !slug.includes("/") &&
+        !slug.includes("\\") &&
+        !slug.includes("#") &&
+        !slug.includes("?")
+      );
+
+    return false;
+  });
+
+  const similarArticles = await articles.reduce(
```

Comment:
Yeah I think that's because those articles have many categories listed and the backend just picks the first (best) one. I can have a go at improving it but it's very slow already

Reply:
Like it's slow on my machine, can't imagine it'll be quick on prod

Reply:
Yes it's quite slow. It took about a minute on my end to generate all the article pages (My DB is about a month old atm). However, the similar articles section is static. So while the build might take some time to complete, the user experience won't be affected.

As for the sad path, if an article has a subcategory but not the category. For example Campus Comment and not Comment, the sad path checks if the category (Campus Comment) is nested under null or 4. Since the category is not nested under either, the sad path goes to step 4. In step 4, get the parent category of Campus Comment using the nest under id. In this case, the id for the Comment Section. Now we have the correct parent. 

Am I missing something here? I'm happy to jump on a call if you wanna go through this in more detail

Reply:
Because the categories are more often wrong than not, all these additional queries will put more load on the server during the built time, regardless. I'm not even sure Node's V8 memory heap can handle this. Memory usage is already quite high when generating article pages. But sure, I'll have a go

Reply:
Thanks dude! Do you know why the memory usage is so high? I'm wondering if it's because we're loading articles in 1000 blocks. Maybe we can try 500?